source/Aconcagua-Quantitative-Analysis/DerivedUnit.class.st

```diff
@@ -0,0 +1,289 @@
+"
+A derived unit is a unit of measurement that is derived from the fundamental (or base) units
+of a particular system. Derived units are formed by combining base units through multiplication,
+ division, or other mathematical operations to measure more complex quantities.
+"
+Class {
+	#name : 'DerivedUnit',
+	#superclass : 'Unit',
+	#instVars : [
+		'exponentsByUnit',
+		'namesOptional'
+	],
+	#category : 'Aconcagua-Quantitative-Analysis',
+	#package : 'Aconcagua-Quantitative-Analysis'
+}
+
+{ #category : 'private' }
+DerivedUnit class >> composedOfAll: exponentsByUnit [
+
+  ^ self composedOfAll: exponentsByUnit nameOptional: Optional unused
+]
+
+{ #category : 'private' }
+DerivedUnit class >> composedOfAll: exponentsByUnit nameOptional: aNamesOptional [
+
+  | nonZeroExponentsByUnit |
+  nonZeroExponentsByUnit := self dictionaryClass new.
+  exponentsByUnit keysAndValuesDo: [ :unit :exponent |
+      [ nonZeroExponentsByUnit at: unit put: exponent ] unless: exponent isZero ].
+  nonZeroExponentsByUnit size = 1 then: [
+      | baseUnit |
+      baseUnit := nonZeroExponentsByUnit keys anyOne.
+      ( nonZeroExponentsByUnit at: baseUnit ) = 1 then: [ ^ baseUnit ]
+    ].
+  ^ self new initializeComposedOfAll: nonZeroExponentsByUnit nameOptional: aNamesOptional
+]
+
+{ #category : 'private' }
+DerivedUnit class >> dictionaryClass [
+
+  ^ OrderedDictionary
+]
+
+{ #category : 'instance creation' }
+DerivedUnit class >> from: baseUnit and: anotherBaseUnit [
+
+  ^ self composedOfAll: ( self dictionaryClass new
+          at: baseUnit put: 1;
+          at: anotherBaseUnit put: 1;
+          yourself )
+]
+
+{ #category : 'instance creation' }
+DerivedUnit class >> from: baseUnit raisedTo: exponent [
+
+  ^ self composedOfAll: ( self dictionaryClass new
+          at: baseUnit put: exponent;
+          yourself )
+]
+
+{ #category : 'comparing' }
+DerivedUnit >> = object [
+
+  ^ self equalityChecker
+      compareWith: [ :base :target |
+          | baseUnits targetUnits |
+          baseUnits := base units.
+          targetUnits := target units.
+          baseUnits size = targetUnits size and: [
+                ( baseUnits includesAll: targetUnits ) and: [ targetUnits includesAll: baseUnits ] ]
+        ];
+      compareWith: [ :base :target |
+          base units allSatisfy: [ :unit |
+                ( base occurrencesOf: unit ) = ( target occurrencesOf: unit ) ] ];
+      checkAgainst: object
+]
+
+{ #category : 'converting' }
+DerivedUnit >> aliasedBy: singularName plural: pluralName symbol: symbol [
```

Comment:
I suggest renaming this (and the initialization) to knownBy:singularNamed:pluralNamed:representedBy:
or knownAs: instead of knownBy: